kivy/uix/bubble.py

```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
-            self_arrow_layout.orientation = 'vertical'
-            self_arrow_layout.cols = 1
-            self_arrow_layout.size_hint = (None, 1)
-            self_arrow_layout.width = self_arrow_img.height
-
-            rotation = -90 if self_arrow_pos[0] == 'l' else 90
-            self._sctr = sctr = Scatter(do_translation=False,
-                                        rotation=rotation,
-                                        do_rotation=False,
-                                        do_scale=False,
-                                        size_hint=(None, None),
-                                        size=(self_arrow_img.size))
-            sctr.add_widget(self_arrow_img)
-
-            if self_arrow_pos[-4:] == '_top':
-                arrow_list = (Widget(size_hint=(1, .07)),
-                              sctr, Widget(size_hint=(1, .3)))
-            elif self_arrow_pos[-4:] == '_mid':
-                arrow_list = (Widget(), sctr, Widget())
-                Clock.schedule_once(self._update_arrow)
-            elif self_arrow_pos[-7:] == '_bottom':
-                arrow_list = (Widget(), Widget(), sctr)
-
-            if self_arrow_pos[0] == 'l':
-                widget_list = (self_arrow_layout, self_content)
+        x, y = pos
+        if not (0 <= x <= self.width and 0 <= y <= self.height):
+            return None
+
+        # the order of the following list defines the side that the arrow
+        # will be attached to in case of ambiguity (same distances)
+        base_layouts_map = [
+            ("bottom_mid", y),
+            ("top_mid", self.height - y),
+            ("left_mid", x),
+            ("right_mid", self.width - x),
+        ]
+        base_layout_key = min(base_layouts_map, key=lambda val: val[1])[0]
+        arrow_layout = list(Bubble.ARROW_LAYOUTS[base_layout_key])
+
+        arrow_width = self._arrow_image.width
+
+        # This function calculates the proper value for pos_hint, i.e., the
+        # arrow texture does not 'overflow' and stays entirely connected to
+        # the side of the content.
+        def calc_x0(x, length):
+            return x * (length - arrow_width) / (length * length)
+
+        if base_layout_key == "bottom_mid":
+            arrow_layout[-1] = {"top": 1.0, "x": calc_x0(x, self.width)}
+        elif base_layout_key == "top_mid":
+            arrow_layout[-1] = {"bottom": 0.0, "x": calc_x0(x, self.width)}
+        elif base_layout_key == "left_mid":
+            arrow_layout[-1] = {"right": 1.0, "y": calc_x0(y, self.height)}
+        elif base_layout_key == "right_mid":
+            arrow_layout[-1] = {"left": 0.0, "y": calc_x0(y, self.height)}
+        return arrow_layout
+
+    def update_content_size(self, instance, value):
+        self.content_size = self.content.size
+
+    def adjust_position(self):
+        if self.limit_to is not None and not self._temporarily_ignore_limits:
+            if self.limit_to is EventLoop.window:
```

Comment:
This part was mainly copied from the old implementation. I use the Bubble widget mainly in a Scatter Plane where this logic does somehow not really make sense. Maybe someone else can verify that this code still "usefull" in other use cases.